ykrt/src/mt.rs

```diff
@@ -281,28 +348,50 @@ impl MTThread {
                         .0
                         .stop_tracing()
                         .unwrap();
-                    let tir_trace = TirTrace::new(&*SIR, &*sir_trace).unwrap();
-                    let ct = TraceCompiler::<I>::compile(tir_trace);
-                    // Now that we have a compiled trace, we need to put a pointer to it inside
-                    // the corresponding Location's pack. To inhibit ExecutableBuffer's Drop,
-                    // we move it into a box and forget it: the pointer in the pack is now the
-                    // sole handle on the trace.
-                    // FIXME: free up the memory when the trace is no longer used.
-                    let ptr = Box::into_raw(Box::new(ct)) as usize;
-
-                    let new_pack = ptr | PHASE_COMPILED;
-                    loc.pack.store(new_pack, Ordering::Release);
+
+                    // Start a compilation thread.
+                    let (snd, rcv) = channel();
+                    let handle = thread::spawn(move || {
+                        let tir_trace = TirTrace::new(&*SIR, &*sir_trace).unwrap();
+                        snd.send(TraceCompiler::<I>::compile(tir_trace)).ok();
+                    });
+                    let ptr = Box::into_raw(Box::new(CompilingTrace { rcv, handle }));
+                    debug_assert_eq!(ptr as usize & PHASE_TAG, 0);
+                    let new_state = ptr as usize | PHASE_COMPILING;
+                    loc.state.store(new_state, Ordering::Release);
+
                     Rc::get_mut(&mut self.inner).unwrap().tracer = None;
                     return None;
                 }
+                PHASE_COMPILING => {
+                    let compiling =
+                        unsafe { Box::from_raw((lp & !PHASE_TAG) as *mut CompilingTrace<I>) };
+                    match compiling.rcv.try_recv() {
+                        Ok(ct) => {
+                            // If we've got a result via the channel then we can ignore any
+                            // subsequent errors from the compiling thread.
+                            compiling.handle.join().ok();
```

Comment:
I don't think it is necessary to wait on the thread to complete. If you drop the `JoinGuard`, the thread will be detached, so joining is neither required nor allowed to prevent a resource leak.

Reply:
I'm not sure if it's strictly necessary or not and the docs aren't really clear. However, I know from other systems that not calling `join` can create problems. [This blog post](https://matklad.github.io/2019/08/23/join-your-threads.html) suggests one should do it in Rust too, though I'm not entirely sure if the reasoning in there is sound or not. I'd be happy if I definitely don't need to do the `join` as it would help simplify the code a bit more, but I feel I need a bit more documentary evidence to be sure :/

Reply:
I now see what you mean by:

> If you drop the JoinGuard, the thread will be detached

it literally calls `pthread_detach` (or equivalent). Which means you're absolutely right and there's no need to manually `join` things. Excellent! So that simplifies things in https://github.com/softdevteam/yk/pull/171/commits/6e03c452b3c4701656b9981fa41a0f77baded7f8 a little bit, which is always nice. Thanks for the pointer!

Reply:
> This blog post suggests one should do it in Rust too, though I'm not entirely sure if the reasoning in there is sound or not.

The first point doesn't really apply here as the thread is known to almost exist anyway. Same for the second point. As for the third point, you are ignoring the panic anyway.